### PR TITLE
fix(ts): parse multiline declarations correctly

### LINKED ISSUE #12773: @babel/parser(ts): incorrect multiline parsing of module declaration
## Bug Report

- [ ] I would like to work on a fix!

**Current behavior**
New line after `module` should not generate `TSModuleDeclaration` 

https://github.com/microsoft/TypeScript/blob/master/tests/baselines/reference/asiPreventsParsingAsAmbientExternalModule01.js

**Input Code**

```js
declare                // this is the identifier 'declare'
module                 // this is the identifier 'module'
"my external module"   // this is just a string
{ }                    // this is a block body
```

**Expected behavior**
I do expect that babel is going to generate 2 identifiers and string and block body, instead of that it generates:
```diff
 ExpressionStatement: Identifier: declare
- TSModuleDeclaration: TSModuleBlock
+ ExpressionStatement: Identifier: module
+ ExpressionStatement
+ ExpressionStatement
```

**Environment**

- Babel version(s): [e.g. v7.12.0]
- Node/npm version: [e.g. Node 12/npm 7]
- OS: [e.g. macOS 10.15.4, Windows 10]
- Monorepo: [e.g. yes/no/Lerna]
- How you are using Babel: [e.g. `webpack`, `rollup`, `parcel`, `babel-register`]

### PR DESCRIPTION
**Pull request by fedeci**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            | Fixes #12773 
| Patch: Bug Fix?          | Yes
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | Yes
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

We check that the next token does not have a previous line break (only if we want to run `next()`). This is necessary to avoid skipping a token that will not be parsed.

### CODE REVIEW

**Review by nicolo-ribaudo — COMMENTED** [review]
Can you also add this test? https://www.typescriptlang.org/play?ssl=1&ssc=1&pln=3&pc=3#code/CYUwxgNghgTiAEBbA9sArhEBYAUAIgDNlk9cBvAXyA

**Comment by armano2** [comment]
is this going to also work with namespaces?

this seem to be parsed as namespace
```ts
declare namespace 
oo 
{ }
```

**Comment by fedeci** [comment]
@armano2 yes it is parsed as a declaration, it needs to be fixed

**Comment by armano2** [comment]
should i raise new ticket for this?

**Comment by nicolo-ribaudo** [comment]
It's not needed, since those are practically two instances of the same bug @fedeci can fix it in this PR

File: packages/babel-parser/test/fixtures/typescript/module-namespace/head-declare/output.json [hunk 1]
```diff
@@ -80,6 +80,99 @@
...
+        "expression": {
+          "type": "StringLiteral",
+          "start":143,"end":148,"loc":{"start":{"line":16,"column":0},"end":{"line":16,"column":5}},
+          "extra": {
+            "rawValue": "baz",
+            "raw": "'baz'"
+          },
+          "value": "baz"
+        }
+      },
```
Comment (fedeci) [inline packages/babel-parser/test/fixtures/typescript/module-namespace/head-declare/output.json:170]: I just discovered this is missing an `ExpressionStatement` for the `module` keyword. It is the same for `namespace` declarations.

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 2]
```diff
@@ -1530,7 +1530,16 @@ export default (superClass: Class<Parser>): Class<Parser> =>
             if (value === "global") {
               return this.tsParseAmbientExternalModuleDeclaration(nany);
             } else {
-              return this.tsParseDeclaration(nany, value, /* next */ true);
+              const oldState = this.state.clone(true);
```
Comment (fedeci) [inline packages/babel-parser/src/plugins/typescript/index.js:1533]: Not 100% happy with this, if there was a way to check that the next token is a line terminator it would be way better.

Comment (JLHwung) [reply packages/babel-parser/src/plugins/typescript/index.js:1533]: We can clone state in `tsTryParseDeclare`. Since it is a try parse method, if any error is thrown in its subroutine, it should be ignored and the parser state should be restored.

Comment (fedeci) [reply packages/babel-parser/src/plugins/typescript/index.js:1533]: Isn't it already cloned in `tsTryParseDeclare` subroutine? Or you just mean `tsTryParse` should be used?
```js
return this.tsTryParse(() =>
  this.tsParseDeclaration(nany, value, /* next */ true),
);
```

Comment (JLHwung) [reply packages/babel-parser/src/plugins/typescript/index.js:1533]: Although it was named `tsTryParseDeclare`, it does not restore state when error is thrown:

https://github.com/babel/babel/blob/759a82abb1f2ea64c03dc9b298aa03950690fa27/packages/babel-parser/src/plugins/typescript/index.js#L1487-L1538

> you just mean tsTryParse should be used?

That could work, just note that `tsTryParseDeclare` does not equal to `tryParse(() => tsParseDeclaration)`: The former addresses `declare ...` parsing, the latter handles general declaration, e.g.`type X = ...`.

Comment (fedeci) [reply packages/babel-parser/src/plugins/typescript/index.js:1533]: Definitely, but I am not sure if the entire `tsTryParseDeclare` method should be wrapped with `tsTryParse` because it is quite expensive. For the moment I will just update the current fix which is working fine.

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/plugins/typescript/index.js]: https://github.com/babel/babel/pull/12771#discussion_r572350118 might be useful for this.

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 3]
```diff
@@ -1617,18 +1617,24 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
-            return this.tsParseModuleOrNamespaceDeclaration(node);
+          if (!next || !this.hasFollowingLineBreak()) {
+            if (next) this.next();
+            if (!this.isLineTerminator()) {
+              if (this.match(tt.string)) {
+                return this.tsParseAmbientExternalModuleDeclaration(node);
+              } else if (this.match(tt.name)) {
+                return this.tsParseModuleOrNamespaceDeclaration(node);
+              }
+            }
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/plugins/typescript/index.js:1628]: Nit: I find it a little hard to follow this logic, with the two different line terminator checks.
Also, if `next` is true we are checking it twice.

```suggestion
          // "module" cannot be followed by a line break.
          if (next) {
            if (this.hasFollowingLineBreak()) break;
            this.next();
          } else if (this.isLineTerminator()) {
            break;
          }

          if (this.match(tt.string)) {
            return this.tsParseAmbientExternalModuleDeclaration(node);
          } else if (this.match(tt.name)) {
            return this.tsParseModuleOrNamespaceDeclaration(node);
          }
```

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 4]
```diff
@@ -1617,18 +1617,24 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
           break;
 
         case "namespace":
-          if (this.tsCheckLineTerminatorAndMatch(tt.name, next)) {
+          if (!next || !this.hasFollowingLineBreak()) {
             if (next) this.next();
-            return this.tsParseModuleOrNamespaceDeclaration(node);
+            if (!this.isLineTerminator() && this.match(tt.name)) {
+              return this.tsParseModuleOrNamespaceDeclaration(node);
+            }
```
Comment (nicolo-ribaudo) [inline packages/babel-parser/src/plugins/typescript/index.js:1637]: Similarly, we can avoid checking the line break twice when `next` is `true` (even if the code _seems_ longer because it's less compact:

```suggestion
          // "namespace" cannot be followed by a line break.
          if (next) {
            if (this.hasFollowingLineBreak()) break;
            this.next();
          } else if (this.isLineTerminator()) {
            break;
          }

          if (this.match(tt.name)) {
            return this.tsParseModuleOrNamespaceDeclaration(node);
          }
```
```

**Review by nicolo-ribaudo — COMMENTED** [review]
I have a comment on the coding style, otherwise the PR looks good!

**Review by nicolo-ribaudo — APPROVED** [review]

**Review by JLHwung — COMMENTED** [review]
Can you add the following cases? All of them should throw
```ts
declare interface
i
{}

declare enum
e
{}

declare abstract
class A {
}

declare type
A = number
```

The idea is that the line break check should be applied universally to the keyword after `declare`.

**Comment by fedeci** [comment]
```ts
// tsc converts this to an enum declaration according to AST explorer, but all of the others require a fix :)
declare enum
e
{}
```

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 5]
```diff
@@ -1644,39 +1643,43 @@ export default (superClass: Class<Parser>): Class<Parser> =>
...
-    tsCheckLineTerminatorAndMatch(tokenType: TokenType, next: boolean) {
-      return (next || this.match(tokenType)) && !this.isLineTerminator();
+    tsCheckLineTerminator(next: boolean) {
+      if (next) {
+        if (this.hasFollowingLineBreak()) return false;
+        this.next();
+      } else if (this.isLineTerminator()) {
+        return false;
+      }
+      return true;
```
Comment (fedeci) [inline packages/babel-parser/src/plugins/typescript/index.js:1682]: Do you prefer?
```suggestion
    tsCheckLineTerminator(next: boolean) {
      if (next) {
        if (this.hasFollowingLineBreak()) return false;
        this.next();
        return true
      }
      return !this.isLineTerminator();
```

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/plugins/typescript/index.js:1682]: I find them equally readable, feel free to do what you prefer.

File: packages/babel-parser/src/plugins/typescript/index.js [hunk 6]
```diff
@@ -1624,14 +1624,13 @@ export default (superClass: Class<Parser>): Class<Parser> =>
       value: string,
       next: boolean,
     ): ?N.Declaration {
+      // no declaration apart from enum can be followed by a line break.
```
Comment (JLHwung) [inline packages/babel-parser/src/plugins/typescript/index.js:1627]: Not related to this PR: I think `enum` allowing line break after is a TS parser bug. We should report it to TypeScript teams.

**Review by JLHwung — APPROVED** [review]

**Review by armano2 — APPROVED** [review]
i executed alignment tests between typescript and your pr, and it looks like its parsed correctly now, thanks 👍

File: packages/babel-parser/src/plugins/typescript/index.js (hunk 6 above)
Comment (armano2) [reply packages/babel-parser/src/plugins/typescript/index.js:1627]: actual reason for this is how js threats those values

javascript has set of reserved keyword https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Lexical_grammar#keywords

`enum` is a `Future reserved keywords` and you can't use it as identifier in valid js code

Comment (JLHwung) [reply packages/babel-parser/src/plugins/typescript/index.js:1627]: Ah makes sense. Concisely `interface` is also a _future reserved word_, but only in strict mode, while `enum` is always reserved.

Comment (fedeci) [reply packages/babel-parser/src/plugins/typescript/index.js:1627]: Should we parse `interface` like we do for `enum` when strict mode is enabled?

Comment (nicolo-ribaudo) [reply packages/babel-parser/src/plugins/typescript/index.js:1627]: No, we should always match the TS behavior: https://www.typescriptlang.org/play?alwaysStrict=false#code/DYUwLgBAlgdmICcBmBDAxiCBeCAmA3AFCGzzLoiECCEhA3gL5A

**Review by nicolo-ribaudo — APPROVED** [review]

**Comment by nicolo-ribaudo** [comment]
Thanks!
